openedx/core/lib/call_stack_manager.py

```diff
@@ -0,0 +1,138 @@
+}
+where frame is a tuple of
+(file path, Line Number, Context)
+"""
+
+import logging
+import traceback
+import re
+import collections
+from django.db.models import Manager
+
+log = logging.getLogger(__name__)
+
+# Module Level variables
+# dictionary which stores call stacks.
+# { "ModelClasses" : [ListOfFrames]}
+# Frames - ('FilePath','LineNumber','Context')
+# ex. {"<class 'courseware.models.StudentModule'>" : [[(file,line number,context),(---,---,---)],
+#                                                     [(file,line number,context),(---,---,---)]]}
+STACK_BOOK = collections.defaultdict(list)
+
+# filter to trickle down call stacks
+EXCLUDE = ['^.*python2.7.*$', '^.*call_stack_manager.*$', '^.*<exec_function>.*$', '^.*exec_code_object.*$']
+REGULAR_EXPS = [re.compile(x) for x in EXCLUDE]
+
+TRACK_FLAG = True
+
+
+def capture_call_stack(current_model):
+    """
+    stores customised call stacks in global dictionary `STACK_BOOK`, and logs it.
+
+    Arguments:
+    current_model - Name of the model class
+    """
+
+    # holds temporary callstack
+    # frame[0][6:-1] -> File name along with path
+    # frame[1][6:] -> Line Number
+    # frame[2][3:] -> Context
+    temp_call_stack = [(frame[0][6:-1],
+                        frame[1][6:],
+                        frame[2][3:])
+                       for frame in [stack.replace("\n", "").strip().split(',') for stack in traceback.format_stack()]
+                       if not any(reg.match(frame[0]) for reg in REGULAR_EXPS)]
+
+    # avoid duplication.
+    if temp_call_stack not in STACK_BOOK[current_model]:
+        STACK_BOOK[current_model].append(temp_call_stack)
+        log.info("logging new call in global stack book, for %s", current_model)
+        log.info(STACK_BOOK)
+
+
+class CallStackMixin(object):
+    """
+    A mixin class for getting call stacks when Save() and Delete() methods are called
+    """
+    def save(self, *args, **kwargs):
+        """
+        Logs before save and overrides respective model API save()
+        """
+        capture_call_stack(str(type(self)))
+        return super(CallStackMixin, self).save(*args, **kwargs)
+
+    def delete(self, *args, **kwargs):
+        """
+        Logs before delete and overrides respective model API delete()
+        """
+        capture_call_stack(str(type(self)))
+        return super(CallStackMixin, self).delete(*args, **kwargs)
+
+
+class CallStackManager(Manager):
+    """
+    gets call stacks of model classes
+    """
+    def get_query_set(self):
+        """
+        overriding the default queryset API methods
+        """
+        capture_call_stack(str(self.model))
+        return super(CallStackManager, self).get_query_set()
+
+
+def donottrack(func):
+    """
+    function decorator which deals with toggling call stacks
+    How to use -
+    1. Just Import following
+        import from openedx.core.lib.call_stack_manager import donottrack
+    :param func: Argument function
+    :return: wrapped function
```

Comment:
We like to use the Napoleon/Google format style for arguments: https://pypi.python.org/pypi/sphinxcontrib-napoleon.

Also, would it be worth allowing the user to disable tracking for specific models?